xarray/core/dataarray.py

```diff
@@ -130,22 +130,59 @@
     T_XarrayOther = TypeVar("T_XarrayOther", bound="DataArray" | Dataset)
 
 
-def _check_coords_dims(shape, coords, dim):
-    sizes = dict(zip(dim, shape, strict=True))
-    for k, v in coords.items():
-        if any(d not in dim for d in v.dims):
+def check_dataarray_coords(
+    shape: tuple[int, ...], coords: Coordinates, dims: tuple[Hashable, ...]
+):
+    """Check that ``coords`` dimension names and sizes do not conflict with
+    array ``shape`` and dimensions ``dims``.
+
+    If a coordinate is associated with an index, the coordinate may have any
+    arbitrary dimension(s) as long as the index's dimensions (i.e., the union of
+    the dimensions of all coordinates associated with this index) intersects the
+    array dimensions.
+
+    If a coordinate has no index, then its dimensions much match (or be a subset
+    of) the array dimensions. Scalar coordinates are also allowed.
+
+    """
+    indexes = coords.xindexes
+    skip_check_coord_name: set[Hashable] = set()
+    skip_check_dim_size: set[Hashable] = set()
+
+    # check dimension names
+    for name, var in coords.items():
+        if name in skip_check_coord_name:
+            continue
+        elif name in indexes:
+            index_dims = indexes.get_all_dims(name)
+            if any(d in dims for d in index_dims):
+                # can safely skip checking index's non-array dimensions
+                # and index's other coordinates since those must be all
+                # included in the dataarray so the index is not corrupted
+                skip_check_coord_name.update(indexes.get_all_coords(name))
+                skip_check_dim_size.update(d for d in index_dims if d not in dims)
+                raise_error = False
+            else:
+                raise_error = True
+        else:
+            raise_error = any(d not in dims for d in var.dims)
+        if raise_error:
             raise ValueError(
-                f"coordinate {k} has dimensions {v.dims}, but these "
+                f"coordinate {name} has dimensions {var.dims}, but these "
                 "are not a subset of the DataArray "
-                f"dimensions {dim}"
+                f"dimensions {dims}"
             )
```

Comment:
I'm having a hard time following the logic in this loop. Could you please separate it into two separate loops:
1. Update the sets of all coords/names to skip
2. Raise any necessary errors